### PR TITLE
Add DltFileSpinner to fetch message for DltBroker

### PR DESCRIPTION
**Pull request by lzhou3957**
Enable another message_queue fetcher from reading dlt log.

### CODE REVIEW

File: dlt/dlt.py
Comment (yen3) [inline dlt/dlt.py:619]: The variable is not easy to understand. I guess that you want to add non-blocking mode in `__iter__` iteration. Please consider a good variable name to control the state.

File: dlt/dlt.py
Comment (yen3) [inline dlt/dlt.py:831]: The common behavior of `__iter__` is blocking mode. I would like suggest you to add another function like `non_blocking_iter` to implement this function. 

It's also a good practice to share the implementation without copying the code directly.

File: dlt/dlt.py
Comment (yen3) [inline dlt/dlt.py:1053]: Please use the same function signature with `py_dlt_client_main_loop`

File: dlt/dlt_broker_handlers.py
Comment (yen3) [inline dlt/dlt_broker_handlers.py:203]: `DLTFileSpinner` copy a lot of code from `DLTMessageHandler`. It's not a good habit. Please think about how to share the same implementation without inheritance.

I don't see any big difference between `DLTFileSpinner` and `DLTMessageHandler` except `self._client` and `py_dlt_client_main_loop`.

There are several ways to refine it. please refactor it.

hint: 

1. Object-Oriented Programming: Dependency Injection, Strategy Pattern and Data Object.
2. Functional Programming: type class (Haskell), Traits (Rust), Concept (C++20), Protocol (Python)

File: dlt/dlt_broker_handlers.py
Comment (yen3) [inline dlt/dlt_broker_handlers.py:228]: It's the same as `DLTMessageHandler`

File: dlt/dlt_broker_handlers.py
Comment (yen3) [inline dlt/dlt_broker_handlers.py:257]: It's the same as `DLTMessageHandler`

File: dlt/dlt_broker_handlers.py
Comment (yen3) [inline dlt/dlt_broker_handlers.py:293]: use Pathlib

File: tests/dlt_broker_from_file_spinner_test.py
Comment (yen3) [inline tests/dlt_broker_from_file_spinner_test.py:90]: please don't sleep in test as possible. It will delay the unit test time.

File: tests/dlt_broker_from_file_spinner_test.py
Comment (yen3) [inline tests/dlt_broker_from_file_spinner_test.py:145]: same here and above

File: dlt/dlt_broker.py
Comment (yen3) [inline dlt/dlt_broker.py:88]: I don't really like this way to initialize an object inside another class. What if we want to another DLTMessageHandler in the feature? add more else if? It also causes the function signature of `__init__` is not easy to understand and maintain.

Please think about how to initialize DLTBroker with external objects.

Hint: Dependency Injection.

File: dlt/dlt_broker.py
Comment (yen3) [inline dlt/dlt_broker.py:219]: same here and below.

**Review by yen3 — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:88]: I was also thinking about it:

class DLTBroker(...dlt_message_handler...):
     self.msg_handler = dlt_message_handler

but this could cause modification to DLTConnector as well:

class DLTConnector(Connector):
    def  start(self, target):
         dlt_message = DLTMessageHandler(...)
         self.broker = DLTBroker(dlt_message )
         
Not only this part, but also the commonality between DLTFileSpinner and DLTMessageHandler, which only have different run(). A common base class is easier for maintainence. But that would also touch the code of DLTConnector part.

I would prefer not to touch anything to DLTConnector at this PR, but with a following ticket.

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:257]: yeap, aware

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:293]: I also considered that, but I dont see any occurance of Path in this code base, so I follow what is used in this code base

**Review by yen3 — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:88]: https://realpython.com/python-multiple-constructors/#providing-multiple-constructors-with-classmethod-in-python

**Review by yen3 — CHANGES_REQUESTED** [review]

File: tests/dlt_broker_from_file_spinner_test.py
Comment (lzhou3957) [reply tests/dlt_broker_from_file_spinner_test.py:90]: sorry, after writing to dlt file, a short sleep is needed, because I need to first ensure that the messages have been properly written there. The totally unittest takes 59.02 seconds

File: dlt/dlt_broker.py
Comment (aigarius) [reply dlt/dlt_broker.py:88]: I would go against modifying DLTConnector (or whatever else is using DLTBroker). DLTBroker has its own responsibilities of initializing and keeping up the connection to the DLT data source using the parameters given to it. It's perfectly fine for it to choos what data source to use.

File: dlt/dlt_broker.py
Comment (aigarius) [inline dlt/dlt_broker.py:108]: I would promote filename as a named, optional parameter into the signature of the init function and update docstring there to describe what happens when it is passed and not the ip_address

File: dlt/dlt_broker.py
Comment (aigarius) [inline dlt/dlt_broker.py:86]: The ip_address is now an optional param in the __init__ signature, so it should have a default value of None

File: dlt/dlt.py
Comment (lzhou3957) [reply dlt/dlt.py:619]: renamed

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:217]: done

File: tests/dlt_broker_from_file_spinner_test.py
Comment (aigarius) [reply tests/dlt_broker_from_file_spinner_test.py:90]: try with 0.1 seconds. should be enough for the IO

File: dlt/dlt_broker.py
Comment (aigarius) [reply dlt/dlt_broker.py:88]: But for two classes that have most of the methods being identical, well - why not simply inherit one from the other and override one method? There is no need to make an abstract common class if there is no complex inheritance structure, it is perfectly fine to just have a functional class and then another functional class that is "the same, but with a small twist" which inherits from first

**Review by aigarius — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker_handlers.py
Comment (aigarius) [reply dlt/dlt_broker_handlers.py:203]: Without introducing too many fancy mechanics, simply inheriting from DLTMessageHandler and overriding the different methods should work, no?

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:86]: Done, and docstring is updated accordingly

File: dlt/dlt_broker_handlers.py
Comment (aigarius) [reply dlt/dlt_broker_handlers.py:293]: does Pathlib actually give any benefits if you are not manipulating file and folder names in massive ways? a simple and straightforward os.path.exists is not going anywhere.

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:88]: Aigars, I'm holding at such ugly state with duplicate code, because working on the base class could touch the part for DLTConnector, because DLTMessageHandler is singly used by DLTConnector.

And dltConnector will get fixed later with https://asc.bmwgroup.net/mgujira/browse/SIT-39098, which could result in a better base class.

so, my plan is:
1. first update: only singly add another functionality for ioc dlt connector, withouting touching the hu dlt connector at all
2. fix https://asc.bmwgroup.net/mgujira/browse/SIT-39098 and work on the common base class.

So, if the behaviour of DLTConnector is unexpectedly changed, then out of the worst case, we just need to revoke the 2nd step.

File: tests/dlt_broker_from_file_spinner_test.py
Comment (mkunzmann) [reply tests/dlt_broker_from_file_spinner_test.py:90]: a call to sync or something similar?

File: dlt/dlt_broker_handlers.py
Comment (aigarius) [reply dlt/dlt_broker_handlers.py:294]: If this is not producing too many debug log messages, then reducing this to 0.1 seconds would be wise. 1 second delay at execution start is human-noticeable, but 0.1 is not really.

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:108]: it could even better for the 2nd step in my plan:

```
class DLTBroker:
    def __init__(
       self,
       message_handler,
       enable_dlt_time=False,
       enable_filter_set_ack=False,
       filter_set_ack_timeout=2.0,
       ignore_filter_set_ack_timeout=False,
    ):
    self.msg_handler = message_handler

# Then for DLTConnector
message_handler = DLTMessageHandler(ip, port, timeout)
broker = DLTBroker(message_handler, ...)

# Then for DLTIOCConnector
message_handler = DLTFileSpinner(dlt_file)
broker = DLTBroker(message_handler, ...)
```

Sure, this will happen by touching DLTConnector part. So, I would leave it to the 2nd step, not now

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:203]: inheriting from DLTMessageHandler, then we have to pass None for these unnecessary variables we dont need to DLTMessageHandler, and we also dont need self._client_connect(). If class A has more than class B needs, then I dont think class A is a good candidate as a base class.

Sure, current approach is ugly, and inheritance from DLTMessageHandler is equally ugly, why not inheritance? well, i simply pick up the ugly way I want, if the other way does not make it better :-p

File: dlt/dlt.py
Comment (lzhou3957) [reply dlt/dlt.py:1053]: just added docstring

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:294]: emm, handle(None) will generate many debug logs, because handle is majorly supposed to log the real cases of receiving valid messages. Handle(None) is there to allow the sucess of dlt_broker.add_context() even when dlt file is not created yet. 

For the phase of no dlt file, it is not worthwhile to call handle(None) too many times.

but sure, if you still feel 1 second is too painful, then tell me again.

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:293]: agree. simple os.path.exists does its simple and easy work here

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:294]: after user calls dlt_broker.add_context(). But when user calls it, we have no idea. And the timeout of add_context is 2 seconds, so calling handle() with 1 second interval is enough before timeout, and also generates less debug logs

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:110]: No need to wait for 2nd step

```python
class DLTBroker:
    @classmethod
    def create_from_file(cis, ...) -> "DLTBroker"
        message_handler = DLTFileSpinner(...)
        return cls(..., message_handler=message_handler)

    @classmethod
    def create_from_ip(cis, ...) -> "DLTBroker":
        message_handler = DLTMessageHandler(...)
       return cls(..., message_handler=message_handler)

    def __init__(
       self,
       enable_dlt_time=False,
       enable_filter_set_ack=False,
       filter_set_ack_timeout=2.0,
       ignore_filter_set_ack_timeout=False,
       message_handler: Optional[DLTMessageHandler] = None,
    ):
        self.message_handler = message_handler or ... # original code
```

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:90]: @lzhou3957 Please remove internal link. It's a public repo.

File: dlt/dlt_broker_handlers.py
Comment (yen3) [reply dlt/dlt_broker_handlers.py:294]: Please wait it with a timeout. Sleeping 1 second unconditionally is not a good behavior.

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:90]: If you really wan to refactor it later. it's ok. Can you show your plan how to change the code?

For me, inheritance is not a good solution for this problem. 

Please refer this article (it's highly recommended to read) and this sentence.
https://hynek.me/articles/python-subclassing-redux/
> [For inheritance] In all cases you sacrifice reading convenience for writing convenience.

**Review by yen3 — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:90]: thanks for your acceptance of leaving it in such mess way. I know it is not easy for you.

but the code refactory is another topic, that means, disagreement on the way of code refactory should not block the merge of this code.

to be honest, i think a common base class is enough for this easy scenario. thanks for sharing the link, but i dont see extra big benefits of using that for this easy case.

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:294]: all time buffer is updated to 0.1. and now the execution time has been decreased to 38.29 seconds

Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:294]: sorry, i cannot get the point of "setting timeout for file existence". what do you mean exactly?

```
while not self.mp_stop_flag.is_set():
  while timeout_condition:
       if file_exist:
           break
```
then, this timeout will block the perception of mp_stop_flag, right?

but i think you cannot recommend something like this, then I bet i probably dont get what you mean. Could you elaborate more about your idea?

File: tests/dlt_broker_from_file_spinner_test.py
Comment (lzhou3957) [reply tests/dlt_broker_from_file_spinner_test.py]: thanks for suggestion. instead of fsync, i just added "file.flush" after file.write, then it works out. Now the execution time has been reduced to 38.29 seconds

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:110]: message_handler: Optional[DLTMessageHandler] could not be achieved as long as DLTFileSpinner has nothing to do with DLTMessageHandler at this moment, right?

Yeap, we could also have it defined without any type enforcement, but then the API is still not perfect until the 2nd ticket.

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:90]: I don't say that I accept your plan

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:110]: so please read the type annotation slide. you can use `Union`

Comment (khiz678) [reply dlt/dlt_broker.py:110]: I will make comments on the below discussion to consolidate both of these threads.

File: dlt/dlt_broker_handlers.py
Comment (khiz678) [reply dlt/dlt_broker_handlers.py:203]: @lzhou3957 you don't need DLTFileSpinner at all. It is a type of DLTMessageHandler. To keep things simple and still maintain OOP principles, you need to create a base DLTMessagehandler and from that you get two DLTMessageHandler classes one which deals with DLTClient and one that deals with DLTFile. The rest of the same code is shared between the two in the base class. Then using the OOP methodology you instantiate the DLTMessagehandler correspondingly based on the arguments passed to DLTBroker. If there's no IP passed but filepath passed, then you use the File based message handler, otherwise you use DLTClient based message handler as it is written already. This allows us future proof design if we get any other kind of message handling, we can do that easily. DLTBroker doesn't need to be changed much as well. You keep backwards compatibility too. Some technical ways to achieve this have been told by @yen3 already. If you still have concerns we can have a discussion about how to best achieve this. Alot of other unnecessary code changes that you have needed to do, will also go away.

File: dlt/dlt_broker_handlers.py
Comment (khiz678) [inline dlt/dlt_broker_handlers.py:302]: I will not review the entire DLTFilespinner since it needs to go away and is not needed. But this part will be relevant in the DLTMessagehandler for files, hence I am commenting here.

Please use the load method correctly with corresponding filters from dlt module to do this, instead of directly calling cDLTFile.

**Review by khiz678 — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:302]: why cannot I use cDLTFile directly? will it hurt me in anyway? if it does not bring any extra benefits of using load(), then it is my decision to use which way I would prefer.

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:203]: sorry, i dont want to touch the other branch for DLTConnector with DLTClient at this moment.
The common base class is already considered to be handled in the second ticket.

File: dlt/dlt.py
Comment (lzhou3957) [reply dlt/dlt.py:831]: where i added "yield None" are not the places of raising error, so if there is error, then it will still be thrown out.
The added "yield None" are meant to unblock  "for message in cDLTFile" in the main loop (secondary loop), and the main loop is used in the for-loop (first loop) inside DLTFileSpinner.run(). If secondary loop is blocked at the moment of no more message from dlt file, then first loop is also blocked and could not handle further actions.

Please review DLTFileSpinner. If you reject to do that, then you cannot get the whole picture.

DLTFileSpinner is another use case of DLTMessageHandler, for such absolute truth, i knew that from the very beginning. And that is also why Wenshi does not like my way of copy-and-paste the majority part of DLTMessageHandler code, in order to achieve "not touching the branch of DLTConnector". I dont mean the ugly state is the end state, I'm aware of what I'm doing, what I have in mind is to split the big and complicated task into two smaller steps:
1. simple append a new functionality for IOC to mtee, without touching DLTConnector from DLTClient
2. code refactory and bridge DLTFileSpinner and DLTMessageHandler together. At this step, only python-dlt needs to be worked on, i dont even need to touch mtee_core, mtee_gen22, and idcevo. Because the used API will stay the same there in the uplevel.

what I want to achieve with such split: in case somebody complains DLTConnector does not work after this big change, then I dont need to revoke all 2 steps, right? because step 1 does not touch DLTConnector with DLTClient at all. I just need to focus on the change made in step2, and the worst case is to revoke step 2 from python-dlt without change mtee_core, mtee_gen22, idcevo, because the added IOC functionality introduced from step 1 stays there.

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:302]: Please review DLTFileSpinner. If you reject to do that, then you cannot get the whole picture.

DLTFileSpinner is another use case of DLTMessageHandler, for such absolute truth, i knew that from the very beginning. And that is also why Wenshi does not like my way of copy-and-paste the majority part of DLTMessageHandler code, in order to achieve "not touching the branch of DLTConnector". I dont mean the ugly state is the end state, I'm aware of what I'm doing, what I have in mind is to split the big and complicated task into two smaller steps:

simple append a new functionality for IOC to mtee, without touching DLTConnector from DLTClient
code refactory and bridge DLTFileSpinner and DLTMessageHandler together. At this step, only python-dlt needs to be worked on, i dont even need to touch mtee_core, mtee_gen22, and idcevo. Because the used API will stay the same there in the uplevel.
what I want to achieve with such split: in case somebody complains DLTConnector does not work after this big change, then I dont need to revoke all 2 steps, right? because step 1 does not touch DLTConnector with DLTClient at all. I just need to focus on the change made in step2, and the worst case is to revoke step 2 from python-dlt without change mtee_core, mtee_gen22, idcevo, because the added IOC functionality introduced from step 1 stays there.

Comment (yen3) [reply dlt/dlt_broker_handlers.py:302]: @lzhou3957 Please remove your comment. There are a lot of internal repo names and projects. Please consider the confidential rules...

File: dlt/dlt_broker_handlers.py
Comment (aigarius) [inline dlt/dlt_broker_handlers.py:278]: If this is in performance hot path, then calling extra functions is a bad idea

File: dlt/dlt_broker.py
Comment (yen3) [reply dlt/dlt_broker.py:110]: After I take a deep look, I know that it's hard to pass an external DLTMessageHandler from outside. I don't like to init an object from different files. I purpose my solution. FYI

First, add initial functions in `dlt_broker_handlers.py`

```python
def _create_dlt_message_handler(
    ip_address,
    port,
    timeout,
    filter_queue,
    message_queue,
    mp_stop_flag,
    dlt_time_value,
    filter_ack_queue,
    **kwargs,
):
    kwargs["ip_address"] = ip_address
    kwargs["port"] = port
    kwargs["timeout"] = kwargs.get("timeout", DLT_CLIENT_TIMEOUT)

    return DLTMessageHandler(
        filter_queue,
        message_queue,
        mp_stop_flag,
        kwargs,
        dlt_time_value=dlt_time_value,
        filter_ack_queue=filter_ack_queue,
    )


def _create_dlt_file_spinner(filename, filter_queue, message_queue, mp_stop_flag, dlt_time_value, filter_ack_queue):
    return DLTFileSpinner(
        filter_queue,
        message_queue,
        mp_stop_flag,
        kwargs.get("filename"),
        dlt_time_value=dlt_time_value,
        filter_ack_queue=filter_ack_queue,
    )


def create_dlt_message_handler(
    ip_address,
    filename,
    filter_queue,
    message_queue,
    mp_stop_flag,
    dlt_time_value,
    filter_ack_queue,
    **kwargs,
) -> Union[DLTMessageHandler | DLTFileSpinner]:
    if ip_address:
        # If ip_address is given, then messages are retrieved from dlt client at run-time
        return _create_dlt_message_handler(
            ip_address,
            filter_queue,
            message_queue,
            mp_stop_flag,
            dlt_time_value,
            filter_ack_queue,
            **kwargs,
        )
    elif filename:
        # If not ip_address is given, then messages are retrieved from the given filename
        # The logs are written to the given filename from another process
        return _create_dlt_file_spinner(
            kwargs["filename"],
            filter_queue,
            message_queue,
            mp_stop_flag,
            dlt_time_value,
            filter_ack_queue,
            **kwargs,
        )

    raise NotImplementedError()
```

Second, change the `self.msg_handler` init function

```python
self.msg_handler = create_dlt_message_handler(
            ip_address,
            kwargs.get("filename"),
            self.filter_queue,
            self.message_queue,
            self.mp_stop_flag,
            self._dlt_time_value,
            self.filter_ack_queue,
            **kwargs,
        )
```

File: dlt/dlt_broker.py
Comment (yen3) [inline dlt/dlt_broker.py:133]: You can add `show_init_info` in DLTMessageHandler and DLTFileSpinner then call it to avoid using the if-else

File: dlt/dlt_broker.py
Comment (yen3) [inline dlt/dlt_broker.py:201]: same here. add a `fail_ack_message` in DLTMessageHandler and DLTFileSpinner and avoid using if-else here.

File: dlt/dlt_broker.py
Comment (yen3) [inline dlt/dlt_broker.py:227]: It's weird. Why DLTMessageHandler does not need to stop?

if it does not need, why not create an empty function in DLTMessageHandler then avoid the if block?

File: dlt/dlt_broker_handlers.py
Comment (yen3) [inline dlt/dlt_broker_handlers.py:408]: why not add

```
def stop():
    pass
```

then eliminate the if-block in DLTBroker stop function?

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:133]: sorry, this is the log for DLTBroker, not for dlt message handler.

File: dlt/dlt_broker.py
Comment (lzhou3957) [reply dlt/dlt_broker.py:201]: i think you miss the responsibility of dlt broker and dlt message handler. Sure, adding DispatcherBase:fail_ack_message() could remove ths if-else in dlt broker, but why?

This failure only means DLTBroker does not receive ack at the point it checks, but it does not mean DispatcherBase fails at all, because ack will be done when next message arrives.

In a word, this is only the assumption of DLTBroker, which does not tell the truth of DispatcherBase and has nothing to do with DispatcherBase

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:408]: already replied in private chat, and prefer to stay it as it is

File: dlt/dlt_broker_handlers.py
Comment (khiz678) [inline dlt/dlt_broker_handlers.py:362]: aaa what? why are you checking if it's set before setting it? :open_mouth:

**Review by khiz678 — CHANGES_REQUESTED** [review]

File: dlt/dlt_broker_handlers.py
Comment (lzhou3957) [reply dlt/dlt_broker_handlers.py:362]: emm, good catch. a "not" is missing...

**Review by khiz678 — APPROVED** [review]

**Pull request merged**
